HIP/hip-513.md

```diff
@@ -0,0 +1,293 @@
+---
+hip: 513
+title: Smart Contract Traceability Extension
+author: Stoyan Panayotov <stoyan.panayotov@limechain.tech>, Mustafa Uzun <mustafa.uzun@limechain.tech>
+type: Standards Track
+category: Service
+needs-council-approval: Yes
+status: Draft
+last-call-date-time:
+created: 2022-06-08
+discussions-to:
+updated:
+requires: 206, 435, 260
+---
+
+## Abstract
+
+This HIP overrides [HIP-260](./hip-260.md) by moving traceability info from `ContractFunctionResult` to sidecar.
+
+Extends traceability with new sidecar records `ContractAction` and `ContractBytecode`.
+
+Proposes additional data placed in sidecar transaction records to enable full
+replay of transactions by mirror nodes and downstream users of Hedera.
+
+## Motivation
+
+Enabling mirror nodes to download traceability info only if they need or want to process it.
+
+Enabling full replay of smart contract service transactions will improve the
+usability, auditability, and debuggability of smart contract service
+transactions. Current facilities are insufficient to inspect "internal"
+transactions executed across contracts, and provide insufficient error messages
+when correcting errors during smart contract development.
+
+## Rationale
+
+[HIP-435](./hip-435.md) allows the creation of sidecar record files containing
+detailed information about transaction executions without bloating the main
+transaction record stream. This allows us to externalise verbose debug information
+about smart contract transactions that was previously expensive to download and
+process every time.
+
+## User stories
+
+Аs a mirror node operator, I want to have the option to download verbose contract operations-related data
+like contract state changes, contract actions, or contract bytecode.
+
+## Specification
+
+We will introduce a couple of sidecar record types, one containing the minimum
+needed data that allows a user to fully replay the transaction and derive the same
+set of results, and the other containing an entire operation-by-operation trace.
+
+### Add ContractStateChanges sidecar record type
+
+Add and populate a new sidecar record of type `ContractStateChanges`. It will
+have a single field - repeated `ContractStateChange`, which will be populated
+whenever a transaction causes contract storage states to change.
+This field will include the changes across all contract storage and not just the
+initial contract in the transaction.
+
+```
+message ContractStateChanges {
+  repeated ContractStateChange contract_state_changes = 1;
+}
+```
+
+For each slot read or written by the smart contract transaction
+a `ContactStateChange` record should be added to the `ContractStateChanges`.
+
+### Add ContractAction sidecar record type
+
+Add and populate a new sidecar record of type `ContractAction`. It will
```

Comment:
Shouldn't there be a `ContractActions` with a repeated `ContractAction`? How would you show multiple method calls for a particular contract call transaction?

Reply:
DONE